### PR TITLE
Reduce time spent setting ssh session envs

### PR DESCRIPTION
**Pull request by rosstimothy**
`tsh` sets a number of environment variables when setting up the users session. Each key value pair is transmitted one at a time in a "env" ssh request, which adds a number of envs * RTT of additional latency per session.

This introduces a new `envs@goteleport.com` request which sets multiple environment variables in a single ssh request, which reduces the amount of time spent setting envs down to the RTT of a single ssh request. In order to ensure backward compat and interoperability with OpenSSH, if the server does not recognize the `envs@goteleport.com` request the ssh client will resort to sending individual "env" requests.

### CODE REVIEW

File: api/observability/tracing/ssh/session.go [hunk 1]
```diff
@@ -75,6 +77,65 @@ func (s *Session) Setenv(ctx context.Context, name, value string) error {
...
+	)
+	defer span.End()
+
+	raw, err := json.Marshal(envs)
+	if err != nil {
+		return err
+	}
+
+	s.wrapper.addContext(ctx, EnvsRequest)
+	ok, err := s.Session.SendRequest(EnvsRequest, true, ssh.Marshal(EnvsReq{Envs: raw}))
```
Comment (codingllama) [inline api/observability/tracing/ssh/session.go:114]:
```suggestion
	ok, err := s.Session.SendRequest(EnvsRequest, true /* wantReply */, ssh.Marshal(EnvsReq{Envs: raw}))
```

File: api/observability/tracing/ssh/session.go [hunk 2]
```diff
@@ -75,6 +77,65 @@ func (s *Session) Setenv(ctx context.Context, name, value string) error {
...
+			semconv.RPCServiceKey.String("ssh.Session"),
+			semconv.RPCMethodKey.String("SendRequest"),
+			semconv.RPCSystemKey.String("ssh"),
+		),
+	)
+	defer span.End()
+
+	raw, err := json.Marshal(envs)
+	if err != nil {
+		return err
```
Comment (codingllama) [inline api/observability/tracing/ssh/session.go:110]:
```suggestion
		return trace.Wrap(err)
```

Same for others?

File: api/observability/tracing/ssh/ssh.go [hunk 3]
```diff
@@ -35,6 +35,10 @@ import (
 )
 
 const (
+	// EnvsRequest sets multiple environment variables that will be applied to any
+	// command executed by Shell or Run.
```
Comment (codingllama) [inline api/observability/tracing/ssh/ssh.go:39]:
```suggestion
	// command executed by Shell or Run.
	// See [EnvsReq] for the corresponding payload.
```

File: api/observability/tracing/ssh/ssh.go [hunk 4]
```diff
@@ -45,6 +49,14 @@ const (
 	instrumentationName = "otelssh"
 )
 
+// EnvsReq contains json marshaled key:value pairs sent as the
+// payload for an EnvsRequest.
```
Comment (codingllama) [inline api/observability/tracing/ssh/ssh.go:53]:
```suggestion
// payload for an [EnvsRequest].
```

File: api/observability/tracing/ssh/ssh.go [hunk 5]
```diff
@@ -45,6 +49,14 @@ const (
...
+type EnvsReq struct {
+	// Envs is a json marshaled map[string]string containing
+	// environment variables.
+	Envs []byte
```
Comment (codingllama) [inline api/observability/tracing/ssh/ssh.go:57]: Just in case, so it doesn't rely on the field name?

```suggestion
	Envs []byte `json:"envs"`
```

Comment (codingllama) [inline api/observability/tracing/ssh/ssh.go:57]: Suggestion: call the field `EnvsRaw` or `EnvsJSON`?

File: lib/client/session.go [hunk 6]
```diff
@@ -225,22 +225,22 @@ func (ns *NodeSession) createServerSession(ctx context.Context) (*tracessh.Sessi
...
 	for key, val := range ns.env {
-		err = sess.Setenv(ctx, key, val)
-		if err != nil {
-			log.Warn(err)
-		}
+		envs[key] = val
+	}
+
+	if err := sess.SetEnvs(ctx, envs); err != nil {
+		log.Warn(err)
```
Comment (codingllama) [inline lib/client/session.go:243]: Should we add a prefix message here, or is the err readable enough on its own?

File: lib/srv/forward/sshserver.go [hunk 7]
```diff
@@ -1318,6 +1321,27 @@ func (s *Server) handleEnv(ctx context.Context, ch ssh.Channel, req *ssh.Request
...
+		return trace.Wrap(err, "failed to parse envs request")
+	}
+
+	var envs map[string]string
+	if err := json.Unmarshal(raw.Envs, &envs); err != nil {
+		return trace.Wrap(err, "failed to unmarshal envs")
+	}
+
+	if err := scx.RemoteSession.SetEnvs(ctx, envs); err != nil {
+		s.log.Debugf("Unable to set environment variables: %v", err)
```
Comment (codingllama) [inline lib/srv/forward/sshserver.go:1339]: Your call:

```suggestion
		s.log..WithError(err).Debug("Unable to set environment variables")
```

File: lib/srv/term.go [hunk 8]
```diff
@@ -693,9 +693,7 @@ func (t *remoteTerminal) prepareRemoteSession(ctx context.Context, session *trac
 		teleport.SSHSessionID:           string(scx.SessionID()),
 	}
 
-	for k, v := range envs {
-		if err := session.Setenv(ctx, k, v); err != nil {
-			t.log.Debugf("Unable to set environment variable: %v: %v", k, v)
-		}
+	if err := session.SetEnvs(ctx, envs); err != nil {
+		t.log.Debugf("Unable to set environment variables: %v", err)
```
Comment (codingllama) [inline lib/srv/term.go:697]:
```suggestion
		t.log.WithError(err).Debug("Unable to set environment variables")
```

File: api/observability/tracing/ssh/client_test.go [hunk 9]
```diff
@@ -262,3 +263,164 @@ func TestNewSession(t *testing.T) {
...
+							case <-ctx.Done():
+								return
+
+							case req := <-reqs:
+								if req == nil {
+									return
+								}
+
+								switch {
+								case i == 0 && req.Type == EnvsRequest:
```
Comment (codingllama) [inline api/observability/tracing/ssh/client_test.go:332]:
```suggestion
								case i == 0 && req.Type == EnvsRequest: // accept 1st EnvsRequest
```

File: api/observability/tracing/ssh/client_test.go [hunk 10]
```diff
@@ -262,3 +263,164 @@ func TestNewSession(t *testing.T) {
...
+										}
+
+										if actual != v {
+											_ = req.Reply(false, []byte(fmt.Sprintf("expected value %s for env %s, got %s", v, k, actual)))
+											return
+										}
+									}
+
+									_ = req.Reply(true, nil)
+								case i == 1 && req.Type == EnvsRequest:
```
Comment (codingllama) [inline api/observability/tracing/ssh/client_test.go:359]:
```suggestion
								case i == 1 && req.Type == EnvsRequest: // reject 2nd EnvsRequest (so we test fallbacks)
```

File: api/observability/tracing/ssh/client_test.go [hunk 11]
```diff
@@ -262,3 +263,164 @@ func TestNewSession(t *testing.T) {
...
+									k, v := processEnvRequest(req)
+									fallback[k] = v
+								case i == 3 && req.Type == "env":
+									k, v := processEnvRequest(req)
+									fallback[k] = v
+								case i == 4 && req.Type == "env":
+									k, v := processEnvRequest(req)
+									fallback[k] = v
+
+									for k, v := range expected {
```
Comment (codingllama) [inline api/observability/tracing/ssh/client_test.go:371]: Unless we get 4 env variable we don't check fallback against expected. Could we make sure we are hitting this scenario in the test?

File: api/observability/tracing/ssh/client_test.go [hunk 12]
```diff
@@ -262,3 +263,164 @@ func TestNewSession(t *testing.T) {
...
+											return
+										}
+
+										if actual != v {
+											_ = req.Reply(false, []byte(fmt.Sprintf("expected value %s for env %s, got %s", v, k, actual)))
+											return
+										}
+									}
+								default:
+									_ = req.Reply(false, []byte(fmt.Sprintf("unexpected ssh request %s on iteration %d", req.Type, i)))
```
Comment (codingllama) [inline api/observability/tracing/ssh/client_test.go:384]: Send to `errChan` here too, so if this happens we are sure to catch it in the test?

**Review by codingllama — COMMENTED** [review]
Nice optimization!

File: api/observability/tracing/ssh/client_test.go [hunk 13]
```diff
@@ -262,3 +264,154 @@ func TestNewSession(t *testing.T) {
...
+
+		envs := map[string]string{}
+		for i := 0; i < len(expected); i++ {
+			select {
+			case env := <-envReqC:
+				envs[env.Name] = env.Value
+			case <-time.After(3 * time.Second):
+				t.Fatalf("time out waiting for env request %d to be processed", i)
+			}
+		}
```
Comment (codingllama) [inline api/observability/tracing/ssh/client_test.go:402]: Suggestion: use a single timer?

```suggestion
		envs := map[string]string{}
		envsTimeout := time.NewTimer(3*time.Second)
		defer envsTimeout.Stop()
		for i := 0; i < len(expected); i++ {
			select {
			case env := <-envReqC:
				envs[env.Name] = env.Value
			case <-envsTimeout.C:
				t.Fatalf("Time out waiting for env request %d to be processed", i)
			}
		}
```

File: api/observability/tracing/ssh/client_test.go [hunk 14]
```diff
@@ -262,3 +264,154 @@ func TestNewSession(t *testing.T) {
...
+
+		for k, v := range expected {
+			actual, ok := envs[k]
+			require.True(t, ok, "expected env %s to be set", k)
+
```
Comment (codingllama) [inline api/observability/tracing/ssh/client_test.go:407]: nit:

```suggestion
```

File: lib/srv/forward/sshserver.go
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/srv/forward/sshserver.go:1]: Thanks.

**Review by codingllama — APPROVED** [review]

**Review by marcoandredinis — APPROVED** [review]
LGMT

Just a question: does this optimization work for direct connections and for webui/connect ssh sessions?

**Pull request merged**